include/ozo/detail/deadline.h

```diff
@@ -80,4 +80,82 @@ class deadline_handler {
 
 };
 
+template <typename Stream, typename Handler, typename Result>
+class io_deadline_handler {
+public:
+    template <typename TimeConstraint>
+    io_deadline_handler (Stream& stream, const TimeConstraint& t, Handler handler)
+    : timer_(ozo::detail::get_operation_timer(stream.get_executor(), t)) {
+        auto allocator = asio::get_associated_allocator(handler);
+        ctx_ = std::allocate_shared<context>(allocator, stream, std::move(handler));
+        timer_.async_wait(timer_handler{ctx_});
+    }
+
+    void operator() (error_code ec, Result result) {
+        if (--ctx_->first_call) {
+            timer_.cancel();
+            ctx_->ec = std::move(ec);
+            ctx_->result = std::move(result);
+            ctx_.reset();
+        } else {
+            auto handler = std::move(ctx_->handler);
+            auto ec = std::move(ctx_->ec);
+            ctx_.reset();
+            handler(std::move(ec), std::move(result));
+        }
+    }
+
+    using executor_type = asio::associated_executor_t<Handler>;
+
+    executor_type get_executor() const noexcept { return asio::get_associated_executor(ctx_->handler);}
+
+    using allocator_type = asio::associated_allocator_t<Handler>;
+
+    allocator_type get_allocator() const noexcept { return asio::get_associated_allocator(ctx_->handler);}
+
+private:
+    using timer_type = typename ozo::detail::operation_timer<typename Stream::executor_type>::type;
+
+    struct context {
+        Stream& stream;
+        Handler handler;
+        Result result;
+        error_code ec;
+        int first_call{2};
```

Comment:
why not first_call is bool? int value with first_call name looks weird and bad understanding.

Reply:
Oops, there should be an atomic.
With an atomic variable, this is the only right way to work.